language/diem-framework/modules/CRSN.move

```diff
@@ -0,0 +1,323 @@
+address 0x1 {
+/// A module implementing conflict-resistant sequence numbers (CRSNs).
+/// The specification, and formal description of the acceptance and rejection
+/// criteria, force expiration and window shifting of CRSNs are described in DIP-168.
+module CRSN {
+    use 0x1::BitVector::{Self, BitVector};
+    use 0x1::DiemVersion;
+    use 0x1::Signer;
+    use 0x1::Errors;
+
+    friend 0x1::DiemAccount;
+
+    /// A CRSN  represents a finite slice or window of an "infinite" bitvector
+    /// starting at zero with a size `k` defined dynamically at the time of
+    /// publication of CRSN resource. The `min_nonce` defines the left-hand
+    /// side of the slice, and the slice's state is held in `slots` and is of size `k`.
+    /// Diagrammatically:
+    /// ```
+    /// 1111...000000100001000000...0100001000000...0000...
+    ///        ^             ...                ^
+    ///        |____..._____slots______...______|
+    ///     min_nonce                       min_nonce + k - 1
+    /// ```
+    struct CRSN has key {
+        min_nonce: u64,
+        size: u64,
+        slots: BitVector,
+    }
+
+    /// No CRSN resource exists
+    const ENO_CRSN: u64 = 0;
+    /// A CRSN resource wasn't expected, but one was found
+    const EHAS_CRSN: u64 = 1;
+    /// The size given to the CRSN at the time of publishing was zero, which is not supported
+    const EZERO_SIZE_CRSN: u64 = 2;
+    /// CRSNs are not supported until after Diem version 3
+    const ECRSN_NOT_SUPPORTED_AT_VERSION: u64 = 3;
+
+
+    /// Publish a DSN under `account`. Cannot already have a DSN published.
+    public(friend) fun publish(account: &signer, min_nonce: u64, size: u64) {
+        assert(!has_crsn(Signer::address_of(account)), Errors::invalid_state(EHAS_CRSN));
+        assert(DiemVersion::get() > 3, Errors::invalid_state(ECRSN_NOT_SUPPORTED_AT_VERSION));
+        assert(size > 0, Errors::invalid_argument(EZERO_SIZE_CRSN));
```

Comment:
Perhaps we should also put a fairly conservative upper limit on size as well? Otherwise, it's a fairly convenient account size DDOS vector that is not (I think) available otherwise. Any we can always adjust the limit upward via a contract upgrade if we need to...

Reply:
Bounding this makes sense. I'm wondering what a good upper bound would be though. I'm thinking 256 might be a good bound, but I wonder if that might be too small of a window? If so maybe 1024?